src/index.js

```diff
@@ -70,115 +69,113 @@ module.exports = class RpcEngine extends SafeEventEmitter {
+          delete res.result
+          // return originalError and response
+          return cb(responseError, res)
+        }
+
+        // return response
+        return cb(err, res)
+      })
   }
 
-  _runMiddleware (req, res, onDone) {
-    // flow
-    async.waterfall([
-      (cb) => this._runMiddlewareDown(req, res, cb),
-      checkForCompletion,
-      (returnHandlers, cb) => this._runReturnHandlersUp(returnHandlers, cb),
-    ], onDone)
-
-    function checkForCompletion ({ isComplete, returnHandlers }, cb) {
-      // fail if not completed
-      if (!('result' in res) && !('error' in res)) {
-        const requestBody = JSON.stringify(req, null, 2)
-        const message = `JsonRpcEngine: Response has no error or result for request:\n${requestBody}`
-        return cb(new EthereumRpcError(
-          ERROR_CODES.rpc.internal, message, req,
-        ))
-      }
-      if (!isComplete) {
-        const requestBody = JSON.stringify(req, null, 2)
-        const message = `JsonRpcEngine: Nothing ended request:\n${requestBody}`
-        return cb(new EthereumRpcError(
-          ERROR_CODES.rpc.internal, message, req,
-        ))
-      }
-      // continue
-      return cb(null, returnHandlers)
+  async _runMiddleware (req, res) {
+
+    const { isComplete, returnHandlers } = await this._runMiddlewareDown(req, res)
+
+    this._checkForCompletion(req, res, isComplete)
+
+    for (const handler of returnHandlers) {
+      await new Promise((resolve, reject) => {
+        handler((err) => (err ? reject(err) : resolve()))
+      })
+    }
+  }
+
+  _checkForCompletion (req, res, isComplete) {
+    // fail if not completed
+    if (!('result' in res) && !('error' in res)) {
+      const requestBody = JSON.stringify(req, null, 2)
+      const message = `JsonRpcEngine: Response has no error or result for request:\n${requestBody}`
+      throw new EthereumRpcError(
+        ERROR_CODES.rpc.internal, message, req,
+      )
+    }
+    if (!isComplete) {
+      const requestBody = JSON.stringify(req, null, 2)
+      const message = `JsonRpcEngine: Nothing ended request:\n${requestBody}`
+      throw new EthereumRpcError(
+        ERROR_CODES.rpc.internal, message, req,
+      )
     }
   }
 
   // walks down stack of middleware
-  _runMiddlewareDown (req, res, onDone) {
+  async _runMiddlewareDown (req, res) {
+
     // for climbing back up the stack
     const allReturnHandlers = []
-    // flag for stack return
+    // flag for early return
     let isComplete = false
 
-    // down stack of middleware, call and collect optional allReturnHandlers
-    async.mapSeries(this._middleware, eachMiddleware, completeRequest)
+    // go down stack of middleware, call and collect optional allReturnHandlers
+    for (const middleware of this._middleware) {
+      if (!isComplete) {
+        await runMiddleware(middleware)
+      }
+    }
+
+    const returnHandlers = allReturnHandlers.reverse()
```

Comment:
Previously these were never returned if an error occurred, but now they are returned.

Reply:
Maybe this is fine, since they still aren't actually used in the error case 🤔

Reply:
Oh, wait, maybe they are. `res.result` isn't being deleted in the error scenario anymore, so `_checkForCompletion` would no longer throw.

Reply:
Hm, per the spec, the return handlers should called in case of `res.error`, as long as `_checkForCompletion` does _not_ throw, which it will only do if `(!res.result && !res.error) || isComplete`. See: https://github.com/MetaMask/json-rpc-engine/blob/remove-async/test/basic.spec.js/#L265

I did add back the `res.result` deletion in the middleware in f919617, but that would never cause `_checkForCompletion` to throw as long as `res.error` was set.

Reply:
Oh I see what you mean. Yeah that makes sense.